<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Selective inference for k-means clustering ‚Ä¢ KmeansInference</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Selective inference for k-means clustering">
<meta property="og:description" content="An implementation for a difference-in-means test between clusters of observations identified via k-means clustering, as described in Chen and Witten (2022+)">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">KmeansInference</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/technical_details.html">Technical details</a>
</li>
<li>
  <a href="articles/Tutorials.html">Software tutorials</a>
</li>
<li>
  <a href="articles/real_data_example.html">Application to single-cell RNA-seq data</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="kmeansinference-">KmeansInference <img src="reference/figures/sticker.png" align="right" width="150px"><a class="anchor" aria-label="anchor" href="#kmeansinference-"></a>
</h1></div>
<div class="section level3">
<h3 id="what-is-kmeansinference">What is KmeansInference?<a class="anchor" aria-label="anchor" href="#what-is-kmeansinference"></a>
</h3>
<p><code>KmeansInference</code> is an <code>R</code> package for testing for a difference in means between clusters of observations identified via k-means clustering.</p>
</div>
<div class="section level3">
<h3 id="how-do-i-install-the-package">How do I install the package?<a class="anchor" aria-label="anchor" href="#how-do-i-install-the-package"></a>
</h3>
<p>To download (and load) the KmeansInference package, use the code below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"yiqunchen/KmeansInference"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">KmeansInference</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="why-do-we-need-kmeansinference">Why do we need KmeansInference?<a class="anchor" aria-label="anchor" href="#why-do-we-need-kmeansinference"></a>
</h3>
<p>Testing for a difference in means between two groups is one of the most fundamental tasks in statistics, with numerous applications. If the groups under investigation are <em>pre-specified</em>, i.e., defined without making use of the observed data, then classical hypothesis tests will control the Type I error rate.</p>
<p>However, in practice, data analysts often find themselves testing the null hypothesis of a difference in means between two groups that are <em>defined through the same data</em> used for testing, e.g., via the output of a clustering algorithm. For instance, in single-cell RNA-sequencing analysis, researchers often first cluster the cells, and then test for a difference in the mean gene expression levels between the clusters to quantify up- or down-regulation of genes and identify new cell types. In this case, because the groups are defined through the data, classical hypothesis tests lead to an <em>inflated Type I error</em>.</p>
<p>In particular, the inferential challenges resulting from testing data-guided hypotheses have been described as a ‚Äúgrand challenge‚Äù in the field of genomics (L√§hnemann et al., 2020), and papers in the field continue to overlook this issue: as an example, seurat (Stuart et al., 2019), the state-of-the-art single-cell RNA sequencing analysis tool, tests for differential gene expression between groups obtained via clustering, with a note that ‚Äú<em>p-values [from these hypotheses] should be interpreted cautiously, as the genes used for clustering are the same genes tested for differential expression.</em>‚Äù</p>
<p>In this work, we propose a test for a difference in means between two clusters estimated from applying <strong>k-means clustering</strong>. Given a set of observations, suppose that we first apply k-means clustering to obtain clusters. We might then consider testing the null hypothesis that the mean is the same across two <em>estimated clusters</em>. As an example, consider a simulated data set where <span class="math inline"><em>x</em><sub>1</sub>,‚ÄÜ‚Ä¶,‚ÄÜ<em>x</em><sub>150</sub>‚ÄÑ‚àº‚ÄÑùí©(0<sub>2</sub>,<strong>I</strong><sub>2</sub>)</span>, and applying k-means clustering with K=3 yields three clusters (see leftmost panel of Figure 1). In the middle panel, we display the quantiles of the naive p-value for testing the null hypothesis that a randomly chosen pair of <em>estimated clusters</em> have the same mean. Because the naive p-value does not account for the fact that the cluster labels are obtained through the same data used for testing, the corresponding test leads to an extremely inflated Type I error. By contrast, our selective p-value, conditional on the event that k-means clustering yields the two estimated clusters under investigation, leads to a correctly-sized test (see the rightmost panel of Figure 1). More detailed simulation results can be found in Section 2 of our manuscript.</p>
<div>
<figure><img src="reference/figures/figure_1_a.png" alt="Illustrative example" width="30%" class="center"><img src="reference/figures/figure_1_b.png" alt="1b" width="30%" class="center"><img src="reference/figures/figure_1_c.png" alt="1c" width="30%" class="center"></figure><!-- [Figure 1: (a): The piecewise mean structure of $\beta$ according to a two-dimensional grid graph. (b): Under the null hypothesis, both $p_{\text{Hyun}}$ and $p_{C_1,C_2}$ control the selective Type I error, but the z-test $p_{\text{Naive}} = \mathbb{P}(|\nu^\top Y|\geq |\nu^\top y|)$ leads to inflated selective Type I error. (c): For a given value of the effect size ($|\nu^\top\beta|/\sigma$), $p_{C_1,C_2}$ has higher power than $p_{\text{Hyun}}$. Power for both increases as a function of the effect size.] -->
</div>
<div class="section level3">
<h3 id="link-to-additional-resources">Link to additional resources<a class="anchor" aria-label="anchor" href="#link-to-additional-resources"></a>
</h3>
<ul>
<li>You can learn more about the technical details in our manuscript and in the <a href="https://yiqunchen.github.io/KmeansInference/articles/technical_details.html" class="external-link">technical details section</a>.</li>
<li>You can learn more about how to use our software in the <a href="https://yiqunchen.github.io/KmeansInference/articles/Tutorials.html" class="external-link">tutorials section</a>.</li>
<li>Finally, code and steps to reproduce the figures in our manuscript can be found in the GitHub repo <a href="https://github.com/yiqunchen/KmeansInference-experiments" class="external-link">https://github.com/yiqunchen/KmeansInference-experiments</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>If you use <code>KmeansInference</code> for your analysis, please cite our manuscript:</p>
<p>Chen YT, Witten DM. (2022+) Selective inference for k-means clustering. arXiv preprint. <a href="https://arxiv.org/abs/xxxx.xxxxx" class="external-link uri">https://arxiv.org/abs/xxxx.xxxxx</a>.</p>
</div>
<div class="section level3">
<h3 id="bug-reports--change-requests">Bug Reports / Change Requests<a class="anchor" aria-label="anchor" href="#bug-reports--change-requests"></a>
</h3>
<p>If you encounter a bug or would like to make a change request, please file it as an issue <a href="https://github.com/yiqunchen/KmeansInference/issues" class="external-link">here</a>.</p>
</div>
<div class="section level3">
<h3 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<p>Chen YT, Witten DM. (2022+) Selective inference for k-means clustering. arXiv preprint. <a href="https://arxiv.org/abs/xxxx.xxxxx" class="external-link uri">https://arxiv.org/abs/xxxx.xxxxx</a>.</p>
<p>Fithian W, Sun D, Taylor J. (2014) Optimal Inference After Model Selection. arXiv:1410.2597 [mathST].</p>
<p>Gao, L. L., Bien, J., and Witten, D. (2022+). Selective inference for hierarchical clustering. arXiv:2012.02936.</p>
<p>L√§hnemann, D., K√∂ster, J., Szczurek, E., et al.¬†(2020). Eleven grand challenges in single-cell data science. Genome Biology, 21(1):31.</p>
<p>Stuart, T., Butler, A., Hoffman, P., et al.¬†(2019). Comprehensive integration of Single-Cell data. Cell, 177(7):1888‚Äì1902.e21.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing KmeansInference</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yiqun Chen <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Yiqun Chen.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
